src/mscorlib/src/System/Collections/Generic/Comparer.cs

```diff
@@ -45,58 +45,84 @@ public static Comparer<T> Create(Comparison<T> comparison)
         //
         private static Comparer<T> CreateComparer()
         {
-            object result = null;
-            RuntimeType t = (RuntimeType)typeof(T);
+            var type = (RuntimeType)typeof(T);
 
             // If T implements IComparable<T> return a GenericComparer<T>
-            if (typeof(IComparable<T>).IsAssignableFrom(t))
+            if (typeof(IComparable<T>).IsAssignableFrom(type))
             {
-                result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(GenericComparer<int>), t);
+                return (Comparer<T>)RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(GenericComparer<int>), type);
             }
-            else if (default(T) == null)
+
+            if (default(T) == null)
             {
-                // If T is a Nullable<U> where U implements IComparable<U> return a NullableComparer<U>
-                if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(Nullable<>)) {
-                    RuntimeType u = (RuntimeType)t.GetGenericArguments()[0];
-                    if (typeof(IComparable<>).MakeGenericType(u).IsAssignableFrom(u)) {
-                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(NullableComparer<int>), u);
-                    }
+                // Nullable does not implement IComparable<T?> directly because that would add an extra interface call per comparison.
+                // Instead, it relies on Comparer<T?>.Default to specialize for nullables and do the lifted comparisons if T implements IComparable.
+                if (type.IsValueType)
```

Comment:
Using `default(T) == null && typeof(T).IsValueType` to detect nullables here, after discussion at https://github.com/dotnet/corefx/issues/16217#issuecomment-280222155